ckan/config/middleware/flask_app.py

```diff
@@ -90,15 +95,47 @@ def _get_interface(self, app: CKANApp):
 
 
 class I18nMiddleware(object):
-    def __init__(self, app: CKANApp):
+    def __init__(self, app: Any):
         self.app = app
 
     def __call__(self, environ: Any, start_response: Any):
-
         handle_i18n(environ)
         return self.app(environ, start_response)
 
 
+def handle_i18n(base_environ: dict[str, Any] | None) -> None:
+    u'''
+    Strips the locale code from the requested url
+    (eg '/sk/about' -> '/about') and sets environ variables for the
+    language selected:
+
+        * CKAN_LANG is the language code eg en, fr
+        * CKAN_LANG_IS_DEFAULT is set to True or False
+        * CKAN_CURRENT_URL is set to the current application url
+    '''
+    environ = base_environ or request.environ
+    locale_list = get_locales_from_config()
+    default_locale = config.get(u'ckan.locale_default')
+
+    # We only update once for a request so we can keep
+    # the language and original url which helps with 404 pages etc
+    if u'CKAN_LANG' not in environ:
+        path_parts = environ[u'PATH_INFO'].split(u'/')
+        if len(path_parts) > 1 and path_parts[1] in locale_list:
```

Comment:
`ckan.root_path` might have a more complicated structure, see the example in https://docs.ckan.org/en/2.11/maintaining/configuration.html#ckan-root-path

Reply:
Yes, makes sense. But this is an unchanged code of the function that I moved into this module because it's used only here by I18n middleware. I'll probably fix it inside simplify-url-for PR as it's more related to locale, than the current PR with authentication fixes.

Reply:
Turned out, it's not a mistake. Whatever root-path is, it never get's into CKAN - `/data/{{LANG}}` or `/x/y/{{LANG}}/z/f` - in both cases when user visits for example organization page in CKAN with such root path, application sees only part after the root-path, i.e `/organization`. 
It means, it's responsibility of the server to insert the language in the beginning of the path when proxifying the request. For example, using following nginx configuration for `/data/{{LANG}}/site` root path:
```ini
    location ~ \/data(\/\w+)?\/site\/(.*) {
        proxy_pass   http://127.0.0.1:5000$1/$2;
    }
```

Reply:
I don't think that's the way we've used it in the past. Where is this documented?

Reply:
There is no documentation that explains how locale gets into CKAN, that's the behavior I see in the few last versions - CKAN does not understands the locale if it's not in the beginning of the URL that gots into CKAN.
Documentation of the [root_path](https://docs.ckan.org/en/2.11/maintaining/configuration.html#ckan-root-path) mentions only URL construction and says nothing about URL parsing, so it seems that that was the behaviour that no one knew about.